SSA/Projects/InstCombine/AliveAutoGenerated.lean

```diff
@@ -1,12 +1,101 @@
 
 infixr:90 " ⊑ "  => ICom.Refinement
 
+/- TODO: is this theorem true? -/
+@[simp]
+theorem Bitvec.ofNat_zero : Bitvec.ofNat w 0 = 0 := by
+  induction w
+  case zero => simp[Bitvec.ofNat]
+  case succ w' wh =>
+    simp[Bitvec.ofNat]
+    rw[wh]
+    simp[OfNat.ofNat, Zero.zero]
+    sorry
+
+@[simp]
+theorem one_by_two_is_zero : 1 / 2 = (0 : Nat) := rfl
+
+@[simp]
+theorem Bitvec.ofNat_one : Bitvec.ofNat w 1 = 1 := by
+  cases w
+  case zero => simp[Bitvec.ofNat]
+  case succ w' =>
+    simp[Bitvec.ofNat]
+    sorry -- to be proven...
+
+
+
+/--
+- We first generalize `v0...vn`, which gives us names for the accessors of the context.
+- Then we `cases` on these, to clear away the case where the variables could be `none`, leaving us with the
+  interesting case where the variable is a `some value`.
+- At this point, the proof state is what we want. However, the names `x0..xn` are *inaccessible* to the
+  user since the names were generated from within the macro.
+- So, we choose to revert the variables `x0...xn` so that the user can introduce them as she likes.
+- If we directly revert `x0...xn`, we wind up getting a proof state of the form
+    `∀ x0 V0 ... xn Vn`, where the `{ Vi }` is pure noise.
+- So we first clear the `V0...Vn` equations, and then revert `x0...xn`.
+- This gives us a statement of the form `∀ x0... xn, GOAL`
+-/
+macro "simp_mlir" : tactic =>
+  `(tactic|
+      (
+      dsimp only[]
+      unfold ICom.Refinement
+      intros Γv
+      try simp only[ICom.denote, IExpr.denote, HVector.denote, OpDenote.denote, InstCombine.Op.denote, HVector.toPair, pairMapM,
+        HVector.toTuple]
+      try simp[bind, Option.bind, pure]
+      try simp[DerivedContext.ofContext, DerivedContext.snoc, Ctxt.snoc] -- cannot rewrite with simp theorem 'motive is not type correct'
+      try simp[MOp.instantiateCom, InstCombine.MTy.instantiate, ConcreteOrMVar.instantiate,
+          Vector.get, HVector.toSingle, HVector.toTuple]
+      try simp[List.nthLe]
+
+      try generalize V0 : Γv (Ctxt.Var.last _ _) = v0
+      try generalize V1 : Γv (Ctxt.Var.last _ _).toSnoc = v1
+      try generalize V2 : Γv (Ctxt.Var.last _ _).toSnoc.toSnoc = v2
+      try generalize V3 : Γv (Ctxt.Var.last _ _).toSnoc.toSnoc.toSnoc = v3
+      try generalize V4 : Γv (Ctxt.Var.last _ _).toSnoc.toSnoc.toSnoc.toSnoc = v4
+      try generalize V5 : Γv (Ctxt.Var.last _ _).toSnoc.toSnoc.toSnoc.toSnoc.toSnoc = v5
+
+      try simp[Goedel.toType] at v0
+      try simp[Goedel.toType] at v1
+      try simp[Goedel.toType] at v2
+      try simp[Goedel.toType] at v3
+      try simp[Goedel.toType] at v4
+      try simp[Goedel.toType] at v5
+      /- cases on variables, eliminating trivial `none` cases. -/
+      try cases' v0 with x0 <;> simp
+      <;> try cases' v1 with x1 <;> simp
+      <;> try cases' v2 with x2 <;> simp
+      <;> try cases' v3 with x3 <;> simp
+      <;> try cases' v4 with x4 <;> simp
+      <;> try cases' v5 with x5 <;> simp
```

Comment:
A lot of this seems to be replicated from `SSA/Core/Framework.lean`. I had hoped that simp_mlir would remain generic and independent of the dialects. Is this not the case?

Reply:
It might be a bit annoying to have separate `simp_mlir` and `simp_alive` tactics, but we can follow a similar strategy as `simp_peephole`: have a core tactic which takes extra simp-lemmas as argument, and then have `simp_alive` simply be an alias for `simp_mlir [alive_simp_lemma_1, alive_simp_lemma_2, ...]`.

We can probably have a single tactic at the base of both `simp_mlir` and `simp_peephole`